core/src/test/java/org/infinispan/tx/EntryWrappingInterceptorDoesNotBlockTest.java

```diff
@@ -0,0 +1,310 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import org.infinispan.Cache;
+import org.infinispan.commands.ReplicableCommand;
+import org.infinispan.commands.remote.ClusteredGetCommand;
+import org.infinispan.commands.tx.PrepareCommand;
+import org.infinispan.commons.api.functional.EntryView;
+import org.infinispan.commons.api.functional.FunctionalMap;
+import org.infinispan.commons.api.functional.Traversable;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.context.Flag;
+import org.infinispan.context.impl.TxInvocationContext;
+import org.infinispan.distribution.MagicKey;
+import org.infinispan.functional.impl.FunctionalMapImpl;
+import org.infinispan.functional.impl.ReadWriteMapImpl;
+import org.infinispan.interceptors.DDAsyncInterceptor;
+import org.infinispan.interceptors.InvocationStage;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.TopologyChanged;
+import org.infinispan.notifications.cachelistener.event.TopologyChangedEvent;
+import org.infinispan.remoting.responses.Response;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.remoting.rpc.RpcOptions;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.statetransfer.StateResponseCommand;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.test.fwk.CleanupAfterMethod;
+import org.infinispan.transaction.TransactionMode;
+import org.infinispan.tx.dld.ControlledRpcManager;
+import org.infinispan.util.AbstractControlledRpcManager;
+import org.infinispan.util.ControlledConsistentHashFactory;
+import org.testng.annotations.Test;
+
+@Test(groups = "functional", testName = "tx.EntryWrappingInterceptorDoesNotBlockTest")
+@CleanupAfterMethod
+public class EntryWrappingInterceptorDoesNotBlockTest extends MultipleCacheManagersTest {
+   ConfigurationBuilder cb;
+   ControlledConsistentHashFactory chFactory;
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      chFactory = new ControlledConsistentHashFactory(new int[] { 0, 1 } , new int[] { 0, 2 });
+      cb = new ConfigurationBuilder();
+      cb.clustering().cacheMode(CacheMode.DIST_SYNC).hash().consistentHashFactory(chFactory).numSegments(2);
+      cb.transaction().transactionMode(TransactionMode.TRANSACTIONAL);
+      createCluster(cb, 3);
+      // make sure the caches are started in proper order
+      for (int i = 0; i < 3; ++i) cache(i);
+   }
+
+   public void test1() throws Exception {
+      testMovingStable(this::readWriteKey);
```

Comment:
How about using a DataProvider, something like this:
https://paste.fedoraproject.org/paste/GwE3F5Cv3sQ9X-V1qwwZp15M1UNdIGYhyRLivL9gydE=